servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CancellableSet.java

```diff
@@ -0,0 +1,122 @@
+import static io.servicetalk.utils.internal.PlatformDependent.throwException;
+import static java.util.Collections.newSetFromMap;
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
+
+final class CancellableSet implements Cancellable {
+    private static final AtomicLongFieldUpdater<CancellableSet> counterUpdater =
+            newUpdater(CancellableSet.class, "counter");
+    private final Set<Cancellable> set;
+    private volatile long counter;
+
+    /**
+     * Create a new instance.
+     */
+    CancellableSet() {
+        this(8);
+    }
+
+    /**
+     * Create a new instance.
+     * @param initialSize The initial size of the internal {@link Set}.
+     */
+    CancellableSet(int initialSize) {
+        set = newSetFromMap(new ConcurrentHashMap<>(initialSize));
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Cancel all {@link Cancellable} that have been previously added via {@link #add(Cancellable)} which have not yet
+     * been cancelled, and all future {@link Cancellable}s added via {@link #add(Cancellable)} will also be cancelled.
+     */
+    @Override
+    public void cancel() {
+        if (counterUpdater.getAndSet(this, Long.MIN_VALUE) >= 0) {
+            Throwable delayedCause = null;
+            for (Cancellable c : set) {
+                try {
+                    // Removal while iterating typically results in ConcurrentModificationException, but not for
+                    // ConcurrentHashMap. We use this approach to avoid concurrent invocation of cancel() between
+                    // this method and add (if they race).
+                    if (set.remove(c)) {
+                        c.cancel();
+                    }
+                } catch (Throwable cause) {
+                    delayedCause = catchUnexpected(delayedCause, cause);
+                }
+            }
+
+            if (delayedCause != null) {
+                throwException(delayedCause);
+            }
+        }
+    }
+
+    /**
+     * Add a {@link Cancellable} that will be cancelled when this object's {@link #cancel()} method is called,
+     * or be cancelled immediately if this object's {@link #cancel()} method has already been called.
+     * @param toAdd The {@link Cancellable} to add.
+     * @return {@code true} if the {@code toAdd} was added, and {@code false} if {@code toAdd} was not added because
+     * it already exists. If {@code false} then {@link Cancellable#cancel()} will be called unless the reason is
+     * {@code toAdd} has already been added.
+     */
+    boolean add(Cancellable toAdd) {
+        if (counter < 0) { // make best effort to avoid the set if cancelled already
+            toAdd.cancel();
+            return false;
+        } else if (!set.add(toAdd)) {
+            return false; // user has implemented equals/hashCode so there is overlap?
+        } else if (counterUpdater.incrementAndGet(this) <= 0) {
```

Comment:
What is the reason for the increment? 

This `counter` feels a lot like a boolean flag to me: A negative value indicates the whole set has been cancelled (or is being cancelled) Why there cannot be a regular volatile read?

Reply:
The increment is an alternative approach to a `CAS` to ensure we observe any state change made by the `cancel` thread. Each thread atomically updating the variable means we can use the result here to guarantee either we have been cancelled and we should clean up, or we have not yet been cancelled (and if cancel happens in the future that thread will observe our previously added element bcz of [JMM happens-before](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4)). I don't see where the JMM provides guarantees around a plain volatile read (e.g. cancel thread sets the value, when is the add thread guaranteed to observe that read). For example this is a sequence where the visibility guarantees aren't clear to me...

```
T1
add(x)
cancellables.offer(x) // true

T2
cancel()
cancelledUpdater.CAS // true
cancelAll() // drains the queue, calls .cancel() on each element, but doesn't observe offer(x)

T1
isCancelled() // false -> memory not yet visible to T1
// x never has .cancel() invoked :(
```

Note the independent volatile read approach was implemented, but [recently changed](https://github.com/apple/servicetalk/commit/f4447d5fc5048edbcb620d4abc0e224525533029#diff-cecd984825c587898fcc3b8f2f68631d1a37702cb53f074e1ee9e0143b2c698c).

Reply:
@Scottmitch: Thank you for taking time to provide the explanation. It made me to refresh my JMM knowledge and there is never enough JMM! I believe there is no difference between volatile read and atomic increment. They are both Synchronization Actions and provide the same visibility guarantees.   

Longer version: 
JMM provides no Wall-Clock guarantee when T1 will read the updated value, but that's irrelevant. Volatile reads and volatile writes are in totally ordered: Synchronization Order. 

Now your example has 3 valid outcomes:
1. T1 will read a positive counter value. This is only possible when T2 has not updated the counter yet. Technically it might have already been updated, but the program must behave as if it was not. Sequential Consistency and all that. This means once T2 starts iterating over the cancellable set it will find the item inserted by T1 and cancel it. 
2. T2 will see a negative value and will try to cancel it on its own
3. Obviously it's still possible both T1 and T2 will try to cancel the item, but you have this case covered. 
4. Other result would be inconsistent with JMM.

Reply:
> Thank you for taking time to provide the explanation. It made me to refresh my JMM knowledge and there is never enough JMM!

No problem. Thanks for your feedback on this, JMM is always a tricky topic and the more eyes the better :)

I haven't forgotten about this but will likely put this on the back burner for a while since it is an optimization. I would like to consult with some additional folks internally and think about it more, I also created a jcstress test to experiment too: https://github.com/Scottmitch/servicetalk-1/tree/jcstress_cancellable_set